### PR TITLE
Alexandria 50 user rating api

### PR DESCRIPTION
**Pull request by JessHua159**
`/api/rating/add` route to add rating to a user
`/api/rating/update` route to update rating for a user
`/api/rating/average` route to get average rating for a user
`/api/rating/ratings` route to get ratings information from a user
`/api/rating/rated` route to get the usernames that a user has rated
`/api/rating/delete` route to delete a rating from a user

### CODE REVIEW

File: src/main/java/com/rpi/alexandria/controller/RatingController.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+@Slf4j
+@FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
+@CrossOrigin("http://localhost:${ui.port}")
+public class RatingController extends BaseController {
+
+	public RatingController(UserService userService) {
+		super(userService);
+	}
+
+	@PostMapping("/add")
```
Comment (anirban-a) [inline src/main/java/com/rpi/alexandria/controller/RatingController.java]: We're following RESTful architecture. So the APIs need to follow the RESTful practices. Change it to:
`@PostMapping("/")`

File: src/main/java/com/rpi/alexandria/controller/RatingController.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+		}
+
+		userService.addRating(loggedInUser, rating);
+
+		AppResponse appResponse = buildAppResponse(String.format("%s successfully added rating of %d to %s",
+				loggedInUserUsername, rating.getRatingValue(), rating.getUsernameOther()), HttpStatus.OK);
+		return new ResponseEntity<>(appResponse, appResponse.getHttpStatus());
+	}
+
+	@PutMapping("/update")
```
Comment (anirban-a) [inline src/main/java/com/rpi/alexandria/controller/RatingController.java]: Same as above: `@PutMapping("/")`

File: src/main/java/com/rpi/alexandria/controller/RatingController.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+			usernamesRatedList.add(usernameRated);
+		}
+
+		AppResponse<List<String>> appResponse = new AppResponse<>(
+				String.format("Retrieved usernames that %s rated", loggedInUser.getUsername()), OffsetDateTime.now(),
+				HttpStatus.OK, "", usernamesRatedList);
+		return new ResponseEntity<>(appResponse, appResponse.getHttpStatus());
+	}
+
+	@DeleteMapping("/delete")
```
Comment (anirban-a) [inline src/main/java/com/rpi/alexandria/controller/RatingController.java]: `@DeleteMapping("/")`

File: src/main/java/com/rpi/alexandria/dto/RatingDTO.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+import lombok.Data;
+
+@Data
+public class RatingDTO {
+
+	String usernameOther;
+
+	int ratingValue;
+
```
Comment (anirban-a) [inline src/main/java/com/rpi/alexandria/dto/RatingDTO.java:11]: need to be `private`

File: src/main/java/com/rpi/alexandria/model/Rating.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import javax.validation.constraints.NotBlank;
+
+@Data
+public class Rating {
+
+	// username of account to add rating to, get rating information from, update rating
+	// of, or delete rating from
+	String usernameOther;
+
+	int ratingValue;
```
Comment (anirban-a) [inline src/main/java/com/rpi/alexandria/model/Rating.java]: need to be `private`

File: src/main/java/com/rpi/alexandria/service/UserService.java [hunk 6]
```diff
@@ -82,6 +81,54 @@ public UserDetails loadUserByUsername(final String username) throws UsernameNotF
...
+	public void deleteRating(User loggedInUser, Rating rating) {
+		String usernameOther = rating.getUsernameOther();
+
+		User otherUser = getUser(usernameOther);
+		otherUser.deleteRating(loggedInUser.getUsername());
+
+		loggedInUser.removeUsernameFromUsernamesRated(usernameOther);
+
+		userRepository.save(otherUser);
+		userRepository.save(loggedInUser);
```
Comment (anirban-a) [inline src/main/java/com/rpi/alexandria/service/UserService.java]: change this to `userRepository.saveAll(List.of(otherUser, loggedInUser));`

**Review by anirban-a — CHANGES_REQUESTED** [review]
Apart from the review comments, also please run `mvn spring-javaformat:apply` before pushing the commits.

**Comment by anirban-a** [comment]
Overall looks good logically, but the code is not following the SOLID principles.

File: src/main/java/com/rpi/alexandria/controller/RatingController.java (hunk 1 above)
Comment (JessHua159) [reply src/main/java/com/rpi/alexandria/controller/RatingController.java]: updated

File: src/main/java/com/rpi/alexandria/controller/RatingController.java (hunk 2 above)
Comment (JessHua159) [reply src/main/java/com/rpi/alexandria/controller/RatingController.java]: updated

File: src/main/java/com/rpi/alexandria/controller/RatingController.java (hunk 3 above)
Comment (JessHua159) [reply src/main/java/com/rpi/alexandria/controller/RatingController.java]: updated

File: src/main/java/com/rpi/alexandria/dto/RatingDTO.java (hunk 4 above)
Comment (JessHua159) [reply src/main/java/com/rpi/alexandria/dto/RatingDTO.java:11]: updated

File: src/main/java/com/rpi/alexandria/service/UserService.java (hunk 6 above)
Comment (JessHua159) [reply src/main/java/com/rpi/alexandria/service/UserService.java]: updated

**Review by anirban-a — APPROVED** [review]

**Review by MMosley502 — APPROVED** [review]

**Pull request merged**